server/schedule/operator/builder.go

```diff
@@ -0,0 +1,596 @@
+			b.execTransferLeader(plan.leader2)
+			kind |= OpLeader
+		}
+		if plan.remove != nil {
+			b.execRemovePeer(plan.remove)
+			kind |= OpRegion
+		}
+	}
+	if b.targetLeader != 0 && b.currentLeader != b.targetLeader {
+		if b.currentPeers.Get(b.targetLeader) != nil {
+			b.execTransferLeader(b.targetLeader)
+			kind |= OpLeader
+		}
+	}
+	if len(b.steps) == 0 {
+		return kind, errors.New("no operator step is built")
+	}
+	return kind, nil
+}
+
+func (b *Builder) execTransferLeader(id uint64) {
+	p := b.currentPeers.Get(id)
+	b.steps = append(b.steps, TransferLeader{FromStore: b.currentLeader, ToStore: p.GetStoreId()})
+	b.currentLeader = p.GetStoreId()
+}
+
+func (b *Builder) execPromoteLearner(p *metapb.Peer) {
+	b.steps = append(b.steps, PromoteLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+	b.currentPeers.Set(&metapb.Peer{Id: p.GetId(), StoreId: p.GetStoreId()})
+	b.toPromote.Delete(p.GetStoreId())
+}
+
+func (b *Builder) execAddPeer(p *metapb.Peer) {
+	if b.isLigthWeight {
+		b.steps = append(b.steps, AddLightLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+	} else {
+		b.steps = append(b.steps, AddLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+	}
+	if !p.GetIsLearner() {
+		b.steps = append(b.steps, PromoteLearner{ToStore: p.GetStoreId(), PeerID: p.GetId()})
+	}
+	b.currentPeers.Set(p)
+	if b.peerAddStep == nil {
+		b.peerAddStep = make(map[uint64]int)
+		b.peerAddStep[p.GetStoreId()] = len(b.steps)
+	}
+	b.toAdd.Delete(p.GetStoreId())
+}
+
+func (b *Builder) execRemovePeer(p *metapb.Peer) {
+	b.steps = append(b.steps, RemovePeer{FromStore: p.GetStoreId()})
+	b.currentPeers.Delete(p.GetStoreId())
+	b.toRemove.Delete(p.GetStoreId())
+}
+
+// check if a peer can become leader.
+func (b *Builder) allowLeader(peer *metapb.Peer) bool {
+	if peer.GetStoreId() == b.currentLeader {
+		return true
+	}
+	if peer.GetIsLearner() {
+		return false
+	}
+	store := b.cluster.GetStore(peer.GetStoreId())
+	if store == nil {
+		return false
+	}
+	stateFilter := filter.StoreStateFilter{ActionScope: "operator-builder", TransferLeader: true}
+	return !stateFilter.Target(b.cluster, store)
+}
+
+// stepPlan is exec step. It can be:
+// 1. Add a peer. `addOrPromote` and `leader1` is set.
+// 2. Promote a learner. `addOrPromote` and `leader1` is set.
+// 3. Remove a peer. `remove` and `leader2` is set.
+// 4. Replace a peer. All fields are set.
+type stepPlan struct {
+	addOrPromote *metapb.Peer
+	leader1      uint64 // leader before add/promote peer
+	remove       *metapb.Peer
+	leader2      uint64 // leader before remove peer
+}
+
+func (p stepPlan) String() string {
+	return fmt.Sprintf("stepPlan{addOrPromote={%s},remove={%s},leader1=%v,leader2=%v}", p.addOrPromote, p.remove, p.leader1, p.leader2)
+}
+
+func (p stepPlan) empty() bool {
+	return p.addOrPromote == nil && p.remove == nil
+}
+
+func (b *Builder) peerPlan() stepPlan {
+	// Replace has the highest priority because it does not change region's
+	// voter/learner count.
```

Comment:
It may reduce learner count.

Reply:
Replace only tries to replace peer with the same role — we can only replace voter with voter and replace learner with a learner.

Reply:
I think line 396-401 is trying to promote learner to voter and remove a voter.

Reply:
oh good point, I really didn't think this situation clearly. 😢

Reply:
fixed.